heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraCLIController.java

```diff
@@ -112,10 +123,122 @@ public void removeContainers(Set<PackingPlan.ContainerPlan> containersToRemove)
+      throw new TopologyUpdateRecoverableException(
+          "empty resource per container query output by Aurora");
+    }
+
+    String jsonStr = stderr.toString();
+    LOG.info(String.format("Resource per container: %s", jsonStr));
+    ObjectMapper mapper = new ObjectMapper();
+    JsonNode actualObj = mapper.readTree(jsonStr);
+    return actualObj;
+  }
+
+  private boolean hasEnoughQuota(JsonNode resource, Integer count)
+      throws JsonProcessingException, IOException {
+    // calc available quota
+    JsonNode quota = askAuroraQuota();
+
+    JsonNode quotaNode = quota.get("quota");
+    long quotaDiskMb = quotaNode.get("diskMb").longValue();
+    double quotaNumCpus = quotaNode.get("numCpus").doubleValue();
+    long quotaRamMb = quotaNode.get("ramMb").longValue();
+
+    JsonNode prodNode = quota.get("prodSharedConsumption");
+    long prodDiskMb = prodNode.get("diskMb").longValue();
+    double prodNumCpus = prodNode.get("numCpus").doubleValue();
+    long prodRamMb = prodNode.get("ramMb").longValue();
+
+    long availableDiskMb = quotaDiskMb - prodDiskMb;
+    double availableNumCpus = quotaNumCpus - prodNumCpus;
+    long availableRamMb = quotaRamMb - prodRamMb;
+
+    // calc res for one container
+    JsonNode containerNode =
+        resource.get(0).get("active").get(0).get("assignedTask").get("task").get("resources");
+    double containerNumCpus = containerNode.get("numCpus").doubleValue();
+    long containerRamMb = containerNode.get("ramMb").longValue();
+    long containerDiskMb = containerNode.get("diskMb").longValue();
+
+    // check if res is enough
+    if (containerNumCpus * count > availableNumCpus) {
+      String msg =
+          String.format("not enough cpu quota: %f cpu per container x %d container > %f available",
+              containerNumCpus, count, availableNumCpus);
+      throw new TopologyUpdateRecoverableException(msg);
+    } else if (containerRamMb * count > availableRamMb) {
+      String msg =
+          String.format("not enough ram quota: %d ram per container x %d container > %d available",
+              containerRamMb, count, availableRamMb);
+      throw new TopologyUpdateRecoverableException(msg);
+    } else if (containerDiskMb * count > availableDiskMb) {
+      String msg = String.format(
+          "not enough disk quota: %d disk per container x %d container > %d available",
+          containerDiskMb, count, availableDiskMb);
+      throw new TopologyUpdateRecoverableException(msg);
+    }
+    return true;
+  }
+
+  private boolean isProd(JsonNode jsonStatus) {
+    String tier = jsonStatus.get(0).get("active").get(0).get("assignedTask").get("task").get("tier")
+        .textValue();
+    if ("preferred".equals(tier)) {
+      return true;
+    }
+    return false;
+  }
+
   @Override
   public Set<Integer> addContainers(Integer count) {
-    //aurora job add <cluster>/<role>/<env>/<name>/<instance_id> <count>
-    //clone instance 0
+    try {
```

Comment:
I think this `check quota` logic should be at the upper level rather than at this level. i.e. in `AuroraScheduler.addContainers(...)`

Reply:
do you mean adding `check quota` to spi/scheduler interface

Reply:
I mean, we can do this 
```
try {
      JsonNode jsonStatus = askAuroraStatus();
     ...
}
```
in `AuroraScheduler.addContainers(...)`